### ISSUE TITLE
Making re-usable elements containing `indentedBlock`

### ISSUE THREAD

**Issue by multimeric** [pyparsing/pyparsing#190]
This ties in with #189 in that it involves creating re-usable elements, and with #170 since `indentedBlock` needs to be rewritten.

Basically `indentedBlock` has a problem with re-usability, because by taking a stack as an input, it becomes tied to the context in which that stack was created, meaning that using it outside of that context is now prevented.

For example, if I want to make a re-usable element that contains an `indentedBlock`, I could do this, which forces all rules that build upon it to share one global stack, which you may not want: 

```python
# a.py
stack = []
block = LineStart().leaveWhitespace() + indentedBlock(
    some + elements
    indentStack=stack,
)
```
```python
# b.py
from a import block

rule = oneOrMore(block)
```

For example, imagine if two different parsers were running concurrently, parsing two independent files. If they shared a global indentation stack, an indent in one file would cause the parser of the second file to start parsing wrongly.

On the other hand, I could parameterise `block` and all parent elements into a function:
```python
# a.py
def block(stack):
    return LineStart().leaveWhitespace() + indentedBlock(
        some + elements
        indentStack=stack,
    )
```
```python
# b.py
from a import block
stack = []

rule = oneOrMore(block(stack)) 
```

However, this introduces a *lot* of extra complexity and detracts from the readability of the code, in particular because all parent elements that use `block` have to pass a stack into it, meaning they also need to be parameterised by a stack, and so they also have to be a function rather than a `ParserElement` themselves. 

Is there a simple solution to this? Perhaps this is something that could be fixed in the `indentedBlock` rewrite?

**Comment by ptmcg** [comment]
I'll try to give this some attention this week. Thanks for all the notes. Very likely that the indentedBlock rewrite will not be in the 3.0.0a1 pre-release, but in the pre-release after that.

**Comment by TMiguelT** [comment]
Still thinking about this. Is it possible to store state for a given *parse* (such as the current indentation level) that isn't shared with other times the same instances performs a parse? This would avoid the need for a global stack. Maybe I could hack this in using closure-abuse.

**Comment by TMiguelT** [comment]
Now that I think about it, it should be possible to attach the stack to the indentedBlock instance, but then it will only be able to track indents and dedents for the block, but not pass on indentation information to elements deeper in the parse tree. Actually this might be sufficient for my use case, as it would let me remove the global stack, but it won't help in situations where the indenting is needed to be accessed by multiple different elements. I can't think of a better solution than a global stack in this situation.

**Comment by ptmcg** [comment]
I'm still giving this thought - it is the one thing left before pushing out the next 3.0.0 pre-release.

**Comment by TMiguelT** [comment]
Just going to brainstorm a little.

It seems like the problem is sharing state (aka parsing context) between levels of the parse tree. Given the current architecture it seems like you could add a `context` variable that's passed down the parse tree in each `_parse()` call, in the same way that the string and location (`instring` and `loc`) are passed down. 

If such a context dictionary were added, you could then have certain parsing elements "watch" certain keys of that dictionary. For example, you define an indent element:

```python
class IndentedBlock(ParseExpression):
    def __init__(expr: ParseExpression, context_key: str = 'indent'):
        self.expr = expr
        # This allows for different sets of concurrent indentation stacks, in case that's ever needed
        self.context_key = context_key

    def parseImpl(self, instring: str, loc: int, ctx: dict):
        indent_stack = ctx[self.context_key]
 
        # Indentation logic here

        return self.expr._parse(instring, loc, ctx)
```

Would this require rewriting every parse function and break every single custom parse function? Yes. Is it worth it? Possibly not

**Comment by TMiguelT** [comment]
My current thinking is that we can implement the `IndentedBlock` element I proposed in #87 as a replacement for the `indentedBlock` method, which should fix a lot of problems people are encountering, but not the stack portability issue. It can go into `2.X`, maybe as a new class, and we deprecate the original `deprecatedBlock`.

Then, in 3.0.0 when you have the chance to add breaking API changes, we can add something like the context that I've proposed here, removing the need for a global stack. We can then also remove the old deprecated `indentedBlock` function while we're at it.

I understand you don't have a lot of time at present, but what do you think of this plan?

**Comment by ptmcg** [comment]
Reviewing this and my old notes and examples tomorrow (holiday here in the US).

**Comment by ptmcg** [comment]
Changes implemented in 3.0's IndentedBlock class
